lib/suseconnect.pm

```diff
@@ -0,0 +1,107 @@
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Register base product & extensions with SUSEConnect
+# Maintainer: Jiawei Sun <jwsun@suse.com>
+
+package suseconnect;
+
+use base Exporter;
+use Exporter;
+
+use strict;
+use testapi;
+
+
+our @EXPORT = qw (suseconnect_registeration command_register);
+
+sub suseconnect_registeration {
+    my $product_version = get_var('VERSION');
+
+    # register base prodcut
+    command_register($product_version);
+    my @scc_addons = split(/,/, get_var('SCC_ADDONS', ''));
+
+    # register extensions without reg code
+    for my $addon (@scc_addons) {
+        next unless grep { $addon eq $_ } qw(asmm contm lgm pcm tcm wsm);
+        command_register($product_version, $addon);
+    }
+    assert_script_run 'zypper lr';
+}
+
+sub command_register {
+    my ($version, $addon, $addon_regcode) = @_;
+    my $arch = get_var("ARCH");
+
+    # extensions product list
+    my %product_list = (
+        'asmm'  => 'sle-module-adv-systems-management',
+        'contm' => 'sle-module-containers',
+        'hpcm'  => 'sle-module-hpc',
+        'lgm'   => 'sle-module-legacy',
+        'pcm'   => 'sle-module-public-cloud',
+        'tcm'   => 'sle-module-toolchain',
+        'wsm'   => 'sle-module-web-scripting',
+    );
+    # precompile regexes
+    my $zypper_conflict = qr/^Choose from above solutions by number[\s\S,]* \[1/m;
+    my $zypper_continue = qr/^Continue\? \[y/m;
+    my $zypper_done     = qr/Run.*to list these programs|^ZYPPER-DONE/m;
+
+    if (!$addon) {
+        my $reg_code            = get_var("SCC_REGCODE");
+        my $addon_debug_version = int $version;
+        $version =~ s/\-SP/./;
+        script_run("(SUSEConnect -p SLES/$version/$arch --regcode $reg_code) | tee /dev/$serialdev", 0);
+        my $out = wait_serial($zypper_conflict, 240);
+
+        #resovle potential conflicit by zypper
+        if ($out =~ $zypper_conflict) {
+            save_screenshot;
+            script_run("(zypper --no-refresh install --no-recommends -t product SLES;echo ZYPPER-DONE) | tee /dev/$serialdev", 0);
+            $out = wait_serial([$zypper_conflict, $zypper_continue, $zypper_done], 240);
+            while ($out) {
+                if ($out =~ $zypper_conflict) {
+                    send_key '1';
+                    send_key 'ret';
+                    save_screenshot;
+                }
+                elsif ($out =~ $zypper_continue) {
+                    send_key 'y';
+                    send_key 'ret';
+                    save_screenshot;
+                }
+                elsif ($out =~ $zypper_done) {
+                    save_screenshot;
+                    last;
+                }
+                $out = wait_serial([$zypper_conflict, $zypper_continue, $zypper_done], 240);
+            }
+        }
+    }
```

Comment:
I think it is no need to extract as a new method, it has different logic with zypper_migration.pm. This part use to resovle conflict by install the new kernel. And it is expect result when use SUSEconnet to register the target system. But in zypper_migration.pm the conflict will record as a failed result.